### PR TITLE
[MLOps 1.5] Autogenerate data logging configs from the registry (helper functions)

### PR DESCRIPTION
**Pull request by dbogunowicz**
A collection of all the necessary (in my opinion) helper functions from which we can quickly build the actual `data_logging_config_from_predefined` function.

This PR is very much TDD-driven, so the unit-tests serve in my opinion as a solid documentation on each of the helper functions.

### CODE REVIEW

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+
+def _loggers_to_config_string(
+    loggers: Dict[str, Optional[Union[str, List[str]]]]
+) -> str:
+    return ("\n").join(_nested_dict_to_lines(loggers, yaml_str_lines=[]))
+
+
+def _nested_dict_to_lines(
+    value: Any, key=None, yaml_str_lines=[], level=0
```
Comment (dbogunowicz) [inline src/deepsparse/loggers/metric_functions/helpers/config_generation.py:65]: I know @bfineran , using levels once again :)
Was easier for me to lay out this helper function this way. I will refactor this function to remove this arg before landing ok?

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+from deepsparse.loggers.config import MetricFunctionConfig
+
+
+_WHITESPACE = "  "
+
+
+def data_logging_config_from_predefined(
+    group_names: Union[str, List[str]],
+    loggers: Dict[str, Optional[Dict[str, Any]]] = {"python": {}},
```
Comment (bfineran) [inline src/deepsparse/loggers/metric_functions/helpers/config_generation.py:31]: been leaving this comment a lot - we cannot use mutable types as function defaults in python as they can have unintended side effects. read more here: https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py (hunk 1 above)
Comment (bfineran) [inline src/deepsparse/loggers/metric_functions/helpers/config_generation.py:65]: same here on mutable default

Comment (bfineran) [reply src/deepsparse/loggers/metric_functions/helpers/config_generation.py:65]: level here is fine - the difference is the scope is limited to a single function with self contained logic in its recursion, can be more specific if you want and rename to `_level` to indicate top level users shouldn't specify

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    raise NotImplementedError()
+
+
+def _loggers_to_config_string(
+    loggers: Dict[str, Optional[Union[str, List[str]]]]
+) -> str:
+    return ("\n").join(_nested_dict_to_lines(loggers, yaml_str_lines=[]))
+
+
+def _nested_dict_to_lines(
```
Comment (bfineran) [inline src/deepsparse/loggers/metric_functions/helpers/config_generation.py:64]: `_nested_dict_to_yaml_str`

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                new_value, new_key, yaml_str_lines, level + 1
+            )
+        else:
+            if isinstance(new_value, list):
+                # we can safely return here, since we are not expecting nested lists
+                yaml_str_lines.append(
+                    f"{new_key}:\n{textwrap.indent(_metric_functions_configs_to_string(new_value), prefix=_WHITESPACE)}"  # noqa E501
+                )
+            else:
+                yaml_str_lines.append(f"{_WHITESPACE*level}{new_key}: {new_value}")
```
Comment (bfineran) [inline src/deepsparse/loggers/metric_functions/helpers/config_generation.py:81]: should use elif to save a level of indentation

```python
elif isinstance(...):
    ...
else:
    yaml_str_lines...

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+) -> List[str]:
+    # converts a nested dictionary to a list of yaml string lines
+    for new_key, new_value in value.items():
+        if isinstance(new_value, dict):
+            yaml_str_lines.append(f"{_WHITESPACE*level}{new_key}:")
+            yaml_str_lines = _nested_dict_to_lines(
+                new_value, new_key, yaml_str_lines, level + 1
+            )
+        else:
+            if isinstance(new_value, list):
```
Comment (bfineran) [inline src/deepsparse/loggers/metric_functions/helpers/config_generation.py:75]: why not use `str_list_to_yaml`?

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py (hunk 2 above)
Comment (bfineran) [reply src/deepsparse/loggers/metric_functions/helpers/config_generation.py:31]: (moving discussion to #867 since this will go away after rebase)

**Review by bfineran — CHANGES_REQUESTED** [review]
LGTM overall, let's move the base branch from `main` to `[feature/damian/config_generation](https://github.com/neuralmagic/deepsparse/tree/feature/damian/config_generation)`

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py (hunk 2 above)
Comment (dbogunowicz) [reply src/deepsparse/loggers/metric_functions/helpers/config_generation.py:31]: Interesting!

Comment (dbogunowicz) [reply src/deepsparse/loggers/metric_functions/helpers/config_generation.py:31]: I actually now see that this does mess up my code.

Comment (dbogunowicz) [reply src/deepsparse/loggers/metric_functions/helpers/config_generation.py:31]: I actually do experience it right now but with a different function (inside the same module):
```
def _nested_dict_to_lines(
    value: Any, key=None, yaml_str_lines=[], _level=0
) -> List[str]:```

such a counterintuitive thing....

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py (hunk 3 above)
Comment (dbogunowicz) [reply src/deepsparse/loggers/metric_functions/helpers/config_generation.py:64]: This is a bit misleading, this function outputs a list of string lines.

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py (hunk 5 above)
Comment (dbogunowicz) [reply src/deepsparse/loggers/metric_functions/helpers/config_generation.py:75]: Well we do. `_metric_functions_configs_to_string()` uses that function.

**Review by bfineran — APPROVED** [review]

**Pull request merged**